subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -5,18 +5,176 @@
+#if defined(CONFIG_MBEDTLS)
+#if !defined(CONFIG_MBEDTLS_CFG_FILE)
+#include "mbedtls/config.h"
+#else
+#include CONFIG_MBEDTLS_CFG_FILE
+#endif /* CONFIG_MBEDTLS_CFG_FILE */
+
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/net_sockets.h>
+#include <mbedtls/x509.h>
+#include <mbedtls/x509_crt.h>
+#include <mbedtls/ssl.h>
+#include <mbedtls/error.h>
+#include <mbedtls/debug.h>
+#endif /* CONFIG_MBEDTLS */
+
+/** TLS context information. */
+struct tls_context {
+	/** Information whether TLS context is used. */
+	bool is_used;
+
+	/** Secure protocol version running on TLS context. */
+	enum net_ip_protocol_secure tls_version;
+
+#if defined(CONFIG_MBEDTLS)
+	/** mbedTLS context. */
+	mbedtls_ssl_context ssl;
+
+	/** mbedTLS configuration. */
+	mbedtls_ssl_config config;
+#endif /* CONFIG_MBEDTLS */
+};
+
+/* A global pool of TLS contexts. */
+static struct tls_context tls_contexts[CONFIG_NET_SOCKETS_TLS_MAX_CONTEXTS];
+
+/* A mutex for protecting TLS context allocation. */
+static struct k_mutex context_lock;
+
+/* Initialize TLS internals. */
+static int tls_init(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	memset(tls_contexts, 0, sizeof(tls_contexts));
+
+	k_mutex_init(&context_lock);
+
+	return 0;
+}
+
+SYS_INIT(tls_init, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+/* Allocate TLS context. */
+static struct tls_context *tls_alloc(void)
+{
+	int i;
+	struct tls_context *tls = NULL;
+
+	k_mutex_lock(&context_lock, K_FOREVER);
+
+	for (i = 0; i < ARRAY_SIZE(tls_contexts); i++) {
+		if (!tls_contexts[i].is_used) {
+			tls = &tls_contexts[i];
+			memset(tls, 0, sizeof(*tls));
+			tls->is_used = true;
+
+			NET_DBG("Allocated TLS context, %p", tls);
+			break;
+		}
+	}
+
+	k_mutex_unlock(&context_lock);
+
+	if (tls) {
+		mbedtls_ssl_init(&tls->ssl);
+		mbedtls_ssl_config_init(&tls->config);
+	} else {
+		NET_WARN("Failed to allocate TLS context");
+	}
+
+	return tls;
+}
+
+/* Release TLS context. */
+static int tls_release(struct tls_context *tls)
+{
+	if (!PART_OF_ARRAY(tls_contexts, tls)) {
+		NET_ERR("Invalid TLS context");
+		return -EINVAL;
+	}
+
+	if (!tls->is_used) {
+		NET_ERR("Deallocating unused TLS context");
+		return -EACCES;
+	}
+
+	mbedtls_ssl_config_free(&tls->config);
+	mbedtls_ssl_free(&tls->ssl);
+
+	k_mutex_lock(&context_lock, K_FOREVER);
+
+	tls->is_used = false;
+
+	k_mutex_unlock(&context_lock);
+
+	return 0;
+}
 
 int ztls_socket(int family, int type, int proto)
 {
-	return zsock_socket(family, type, proto);
+	enum net_ip_protocol_secure tls_proto = 0;
+	int sock, ret, err;
+
+	if (proto  >= IPPROTO_TLS_1_0 && proto <= IPPROTO_TLS_1_2) {
+		tls_proto = proto;
+		proto = (type == SOCK_STREAM) ? IPPROTO_TCP : IPPROTO_UDP;
+	}
+
+	sock = zsock_socket(family, type, proto);
+	if (sock < 0) {
+		/* errno will be propagated */
+		return -1;
+	}
+
+	if (tls_proto != 0) {
+		/* If TLS protocol is used, allocate TLS context */
+		struct net_context *context = INT_TO_POINTER(sock);
+
+		context->tls = tls_alloc();
```

Comment:
Thank you for looking into this. Let me address the comments inline.

> In a secure socket offloading case, do we need to allocate a TLS context? Maybe (context->tls == NULL) can be the switch to fall back to the non-TLS zsock_ APIs.
Also, there need not be any calls to mbedtls APIs in that case.

With TLS offloading TLS context does not need to be allocated, as the context contains mostly mbedTLS data. As for the fallback, that's already implemented. We only allocate TLS context when secure socket is created (i. e. proto type is TLS, for instance `IPPROTO_TLS_1_2`). When TLS context is not allocated, we do fallback to regular sockets.

> There is a CONFIG_NET_OFFLOAD macro which, if not defined, could compile out these TLS calls from sockets.
> However, some wifi drivers may not actually handle secure socket offload, so we may need to introduce a new definition (eg: CONFIG_SECURE_SOCKET_OFFLOAD) to compile out the TLS and mbedtls calls from the socket layer.

Look's like we'd rather need another switch, `CONFIG_SECURE_SOCKET_OFFLOAD`, just as you mention. At socket shim layer, it would be enough to block TLS context allocation in case we do support TLS offloading.  
net_context would need to accept new protocol types though, so that they can be passed to the offloaded driver. In this solution, we use for instance `IPPROTO_TLS_1_2` proto type to create a secure socket. Currently, this information is kept at the socket shim layer though and net_context is not aware of it.

> setsockopt() calls setting up the socket for TLS, which calls get passed down to net_context_set_opt() - TBD, which offloads to the driver.

This will be needed as well, for socket configuration. We need a way to set certs/keps/ciphers, and that'd be a job for socket options.

> Is this how we see secure socket offload working?

Well, I guess that we have consensus here, that it is doable with `CONFIG_NET_OFFLOAD`.

But from Nordic's point of view though, a better solution would be something like proposal from #4821. I saw that PR got closed due to lack of consensus and socket support in application protocols. But as Zephyr's networking subsystem is undergoing transition to socket API being default, and there are more parites interested in the solution (Nordic), perhaps it would be a good idea to bring back the topic? Forcing the transistion between sockets<->net_context<->sockets sounds like a unnecessary overcomplexity to me.

Reply:
Thanks for the replies...

> net_context would need to accept new protocol types though, so that they can be passed to the offloaded driver. In this solution, we use for instance IPPROTO_TLS_1_2 proto type to create a secure socket. 

Yes, either net_context_get() could allow a new protocol number (non-IANA standard) in `enum net_ip_protocol`, or we could pass another option via setsockopt() to promote an open non-secure socket to a secure socket.  (The SimpleLink stack can handle either).  It may be nice to support both, eventually (if there's a use case).

> But from Nordic's point of view though, a better solution would be something like proposal from #4821. 

As I understand it, the main blockers for that PR were:
1.  It bypasses the net_context APIs, so existing protocols built on net_context do not benefit from the socket offload;
  * This seems to be getting resolved via the plan to move protocols to sockets, and perhaps by the work to split up net_app into libraries.
2.  This socket offload method bypasses the Zephyr IP routing tables, so only one network interface is ever bound to the sockets.   
  * So, for example, packets cannot be routed between network interfaces.  Eg: no multi-bearer support (https://en.wikipedia.org/wiki/Multi-bearer_network).
  * Any other potential uses of the routing table would also be bypassed.  Eg: to bind a socket to a particular network interface.
  * This actually is still OK for our customer's typical use case: an IoT edge device connected via a single network interface (WiFi device).  But, it's a constraint that would need to be known/accepted building for a pure "socket offload" device.

> perhaps it would be a good idea to bring back the topic? 

Given there is interest from at least two important vendors of IoT devices (TI, Nordic), and one of the major issues is getting resolved (protocols migrating to sockets), it seems it may be time to do so.